### PR TITLE
Feature/authentication based routing

### LINKED ISSUE #68: Authentication-based routing
This becomes relevant once we have more non-authentication routes.

- The user should be brought back to the page they were on when going through the logging flow.
- Route guards provide better UX.

### PR DESCRIPTION
**Pull request by XanderVertegaal**
Closes #68 

This PR adds a LoggedOn route guard and handles post-login redirects.

For testing, I've created a separate branch (based on this one) that adds two extra routes, one ('secret') with the route guard and one ('other') without. You can find it by checking out `feature/authentication-based-routing-test` (https://github.com/CentreForDigitalHumanities/lettercraft/tree/feature/authentication-based-routing-test).

### CODE REVIEW

File: frontend/src/app/core/menu/user-menu/user-menu.component.html [hunk 1]
```diff
@@ -22,7 +22,12 @@
     </ng-container>
     <ng-container *ngIf="showSignIn$ | async">
         <li class="nav-item">
-            <a [routerLink]="['/login']" routerLinkActive="active" class="nav-link">
+            <a
+                [routerLink]="['/login']"
+                routerLinkActive="active"
+                [queryParams]="(onLoginPage$ | async) ? null : { next: currentPath$ | async }"
```
Comment (XanderVertegaal) [inline frontend/src/app/core/menu/user-menu/user-menu.component.html:28]: Prevents endless stacking of `?next=?next=?next=...` if a user (for some reason) keeps clicking on `Sign In`.

File: frontend/src/app/core/menu/user-menu/user-menu.component.ts [hunk 2]
```diff
@@ -19,22 +20,24 @@ export class UserMenuComponent implements OnInit {
...
+    public showSignIn$ = this.authService.currentUser$.pipe(map(_.isNull));
 
     public icons = {
         user: faUser,
     };
 
     public logoutLoading$ = this.authService.logout.loading$;
 
+    public currentPath$ = this.currentPath.path$;
+    public onLoginPage$ = this.currentPath.onLoginPage$;
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply frontend/src/app/core/menu/user-menu/user-menu.component.ts:32]: Much easier this way, thanks for the suggestions! 👍

[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply frontend/src/app/core/menu/user-menu/user-menu.component.ts]: In the end, using a getter gave me an ExpressionChangedAfterChecked error, so I stuck to using the Observable that `this.router.routerState.root.url` gave me.

While reading the documentation of `RouterLinkActive`, I found a readonly `isActive` boolean that can also be used to check whether a route is active, making `onLoginPage$` wholly unnecessary.

File: frontend/src/app/core/menu/user-menu/user-menu.component.html [hunk 3]
```diff
@@ -22,7 +22,13 @@
     </ng-container>
     <ng-container *ngIf="showSignIn$ | async">
         <li class="nav-item">
-            <a [routerLink]="['/login']" routerLinkActive="active" class="nav-link">
+            <a
+                #linkActive="routerLinkActive"
+                [routerLink]="['/login']"
+                routerLinkActive="active"
+                [queryParams]="linkActive.isActive ? null : { next: currentPath$ | async }"
```
Comment (lukavdplas) [inline frontend/src/app/core/menu/user-menu/user-menu.component.html:29]: neat solution to use a template variable here! :)

**Review by lukavdplas — APPROVED** [review]

**Pull request merged**